File: sdk-python/src/agent_did_sdk/core/signer.py
"""AgentSigner protocol and LocalKeySigner implementation."""

from __future__ import annotations

from typing import Protocol, runtime_checkable

from nacl.signing import SigningKey


@runtime_checkable
class AgentSigner(Protocol):
    """Interface for signing operations.

    Implement this to integrate with external key management
    (KMS, HSM, hardware wallets) in production.
    """

    async def sign(self, payload: bytes) -> str:
        """Sign *payload* and return the signature as a hex string."""
        ...

    async def get_public_key(self) -> bytes:
        """Return the raw Ed25519 public key (32 bytes)."""
        ...


class LocalKeySigner:
    """Default signer using a local Ed25519 private key.

    Suitable for development, testing, and demo scenarios.
    """

    def __init__(self, private_key_hex: str) -> None:
        self._signing_key = SigningKey(bytes.fromhex(private_key_hex))

    async def sign(self, payload: bytes) -> str:
        signed = self._signing_key.sign(payload)
        return signed.signature.hex()

    async def get_public_key(self) -> bytes:
        return bytes(self._signing_key.verify_key)

    @staticmethod
    def generate() -> tuple["LocalKeySigner", str]:
        """Generate a new signer with a random key. Returns (signer, private_key_hex)."""
        sk = SigningKey.generate()
        hex_key = sk.encode().hex()
        return LocalKeySigner(hex_key), hex_key

File: sdk/src/core/signer.ts
import { ed25519 } from '@noble/curves/ed25519';
import { bytesToHex, hexToBytes } from '@noble/hashes/utils';

/**
 * Interface for signing operations. Implement this to integrate with
 * external key management (KMS, HSM, hardware wallets) in production.
 */
export interface AgentSigner {
  /** Sign a message and return the signature as hex string. */
  sign(payload: Uint8Array): Promise<string>;
  /** Return the raw Ed25519 public key (32 bytes). */
  getPublicKey(): Promise<Uint8Array>;
}

/**
 * Default signer that uses a local Ed25519 private key.
 * Suitable for development, testing, and demo scenarios.
 */
export class LocalKeySigner implements AgentSigner {
  private readonly privateKeyBytes: Uint8Array;

  constructor(privateKeyHex: string) {
    this.privateKeyBytes = hexToBytes(privateKeyHex);
  }

  async sign(payload: Uint8Array): Promise<string> {
    const signatureBytes = ed25519.sign(payload, this.privateKeyBytes);
    return bytesToHex(signatureBytes);
  }

  async getPublicKey(): Promise<Uint8Array> {
    return ed25519.getPublicKey(this.privateKeyBytes);
  }

  /** Generate a new LocalKeySigner with a random key. Returns [signer, privateKeyHex]. */
  static generate(): [LocalKeySigner, string] {
    const privateKeyBytes = ed25519.utils.randomPrivateKey();
    const privateKeyHex = bytesToHex(privateKeyBytes);
    return [new LocalKeySigner(privateKeyHex), privateKeyHex];
  }
}
